// ngrx-data/esm5/entity-metadata/entity-definition.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { createEntityAdapter } from '@ngrx/entity';
import { defaultSelectId } from '../utils/utilities';
/**
 * @record
 * @template T
 */
export function EntityDefinition() { }
/** @type {?} */
EntityDefinition.prototype.entityName;
/** @type {?} */
EntityDefinition.prototype.entityAdapter;
/** @type {?|undefined} */
EntityDefinition.prototype.entityDispatcherOptions;
/** @type {?} */
EntityDefinition.prototype.initialState;
/** @type {?} */
EntityDefinition.prototype.metadata;
/** @type {?} */
EntityDefinition.prototype.noChangeTracking;
/** @type {?} */
EntityDefinition.prototype.selectId;
/** @type {?} */
EntityDefinition.prototype.sortComparer;
/**
 * @template T, S
 * @param {?} metadata
 * @return {?}
 */
export function createEntityDefinition(metadata) {
    /** @type {?} */
    var entityName = metadata.entityName;
    if (!entityName) {
        throw new Error('Missing required entityName');
    }
    metadata.entityName = entityName = entityName.trim();
    /** @type {?} */
    var selectId = metadata.selectId || defaultSelectId;
    /** @type {?} */
    var sortComparer = (metadata.sortComparer = metadata.sortComparer || false);
    /** @type {?} */
    var entityAdapter = createEntityAdapter({ selectId: selectId, sortComparer: sortComparer });
    /** @type {?} */
    var entityDispatcherOptions = metadata.entityDispatcherOptions || {};
    /** @type {?} */
    var initialState = entityAdapter.getInitialState(tslib_1.__assign({ entityName: entityName, filter: '', loaded: false, loading: false, changeState: {} }, (metadata.additionalCollectionState || {})));
    /** @type {?} */
    var noChangeTracking = metadata.noChangeTracking === true; // false by default
    return {
        entityName: entityName,
        entityAdapter: entityAdapter,
        entityDispatcherOptions: entityDispatcherOptions,
        initialState: initialState,
        metadata: metadata,
        noChangeTracking: noChangeTracking,
        selectId: selectId,
        sortComparer: sortComparer
    };
}
