// ngrx-data/esm5/utils/ngrx-entity-models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// Copied from `@ngrx/entity/models` because that lib doesn't export them
// They should be exported by @ngrx/entity
/** @typedef {?} */
var ComparerStr;
export { ComparerStr };
/** @typedef {?} */
var ComparerNum;
export { ComparerNum };
/** @typedef {?} */
var Comparer;
export { Comparer };
/** @typedef {?} */
var IdSelectorStr;
export { IdSelectorStr };
/** @typedef {?} */
var IdSelectorNum;
export { IdSelectorNum };
/** @typedef {?} */
var IdSelector;
export { IdSelector };
/**
 * @record
 * @template T
 */
export function DictionaryNum() { }
/**
 * @abstract
 * @template T
 */
var /**
 * @abstract
 * @template T
 */
Dictionary = /** @class */ (function () {
    function Dictionary() {
    }
    return Dictionary;
}());
/**
 * @abstract
 * @template T
 */
export { Dictionary };
/**
 * @record
 * @template T
 */
export function UpdateStr() { }
/** @type {?} */
UpdateStr.prototype.id;
/** @type {?} */
UpdateStr.prototype.changes;
/**
 * @record
 * @template T
 */
export function UpdateNum() { }
/** @type {?} */
UpdateNum.prototype.id;
/** @type {?} */
UpdateNum.prototype.changes;
/** @typedef {?} */
var Update;
export { Update };
